### DISCUSSION TITLE
serve wrangler as HTTPS

### DISCUSSION THREAD

**Discussion by jamaluddinrumi** [nitrojs/nitro#1807] [General]
i want to run nitro based api server which accepts HTTPS 

![Screenshot 2023-10-09 at 01 08 39](https://github.com/unjs/nitro/assets/446780/43934129-5d43-4128-9640-8074c688a850)

i'm totally confused with [devProxy](https://nitro.unjs.io/config#devproxy) config

for now, i'm ended up using [httpxy](https://github.com/unjs/httpxy) as my https proxy server

```javascript
import fs from "fs";
import { createProxyServer } from "httpxy";

const proxyServer = createProxyServer({
    target: {
        host: 'localhost',
        port: 8282
      },
      ssl: {
        key: fs.readFileSync('127.0.0.1+7-key.pem', 'utf8'),
        cert: fs.readFileSync('127.0.0.1+7.pem', 'utf8')
      }
});

proxyServer.listen(4343)
```

any better ideas of [devProxy](https://nitro.unjs.io/config#devproxy) uses?

**Comment by pi0** [comment]
Hi. Can you please explain better why you need to use `wrangler` serving https? Do you want to run development or production server in Node.js via HTTPS protocol?

**Reply by jamaluddinrumi** [reply]
shortly, all of these is to support running `astro dev` with web worker (by [partytown](https://docs.astro.build/en/guides/integrations-guide/partytown/) ) with HTTPS enabled

i fetch Json API from nitro based via web worker which is also requiring in HTTPS mode 

this is my current topology which works for now
|    	| ubuntu dev server                                                                          	| mac mini   	|
|----	|-------------------------------------------------------------------------------------	|---------------	|
| IP 	| 192.168.8.127                                                                         	| 192.168.8.165 	|
|    	| `astro dev` listening https://192.168.8.127:3232                                      	|               	|
|    	| `nitro` listening http://192.168.8.127:8282                                           	|               	|
|    	| `httpxy` listening https://192.168.8.127:4343 & forwarding to http://192.168.8.127:8282 	|               	|

the main problem is i don't know how to make `nitro` is able to listen the HTTPS version which is https://192.168.8.127:8282

**Reply by jamaluddinrumi** [reply]
i use `mkcert` ssl self certificate generated by my dev server

and already import them to my mac mini

**Reply by jamaluddinrumi** [reply]
_Hi. Can you please explain better why you need to use wrangler serving https?_ 

i run nitro with cloudflare preset 

_Do you want to run development or production server in Node.js via HTTPS protocol?_

yes, i want to run `nitro` with cloudflare preset via HTTPS protocol

**Reply by pi0** [reply]
Have you tried `npx nitro dev --https`? It runs your nitro dev server with SSL and even automatically self-signed cert.

<img width="482" alt="image" src="https://github.com/unjs/nitro/assets/5158436/c1eaaba7-c08d-479f-a58a-08ba66c38da8">

The nitro preset is configured for production deployment only and i guess in your CF worker environment we have already this.

If you are thinking to run "build preview" with cloudflare preset and test it "locally" with wrangler, i think we can add same CLI experience of `nitro dev --https` via `npx httpxy --https --target` (if this is what you want, i just made an issue to make it https://github.com/unjs/httpxy/issues/19)

**Reply by jamaluddinrumi** [reply]
_Have you tried npx nitro dev --https?_ 

yes, i've already tried it

![Screenshot 2023-10-09 at 01 08 39](https://github.com/unjs/nitro/assets/446780/032f406b-5235-45f4-a431-3994a8b11c16)

_It runs your nitro dev server with SSL and even automatically self-signed cert._

yes, it's run with https but is it really self-signed cert automatically? because i get `net::ERR_CERT_AUTHORITY_INVALID` in browser visited from my mac mini 

is this really self-signed cert automatically? because when i run `npx nitro dev --https` it doesn't supply key & cert files 

whereas by using proxy server approach, i supplied those files

![chrome-error___chromewebdata_ (1)](https://github.com/unjs/nitro/assets/446780/c80ddcbe-3f83-409c-93a8-f007c6420457)

**Reply by jamaluddinrumi** [reply]
_If you are thinking to run "build preview" with cloudflare preset and test it "locally" with wrangler, i think we can add same CLI experience of nitro dev --https via npx httpxy --https --target (if this is what you want, i just made an issue to make it https://github.com/unjs/httpxy/issues/19)_

it looks confusing for a newbie like me, 
i thought the spirit of `unjs` is simple-use as possible 

wait, we are out of `devProxy` discussion here? what the use of it? 
is `devProxy` really can provide more simple solution for lets say like-my-case above?

**Reply by jamaluddinrumi** [reply]
the most i expected solution here at least we can supply those self-signed certificate `.pem`s files and running `nitro` with `--https` param aaaand voila it works

**Reply by pi0** [reply]
You can use `--https --https.cert <path_to_cert> --https.key <path_to_key>` to provide your own SSL keys.

`devProxy` option can be used to make requests "from" dev server to outside whereas you want to run your dev server itself with HTTPs enabled.

**Reply by jamaluddinrumi** [reply]
that's what i'm looking for @pi0 

that should be written in docs totally

finally i don't need to use proxy server anymore

**Reply by jamaluddinrumi** [reply]
i forgot to say thank you @pi0

**Reply by pi0** [reply]
Happy helped. PR is more than welcome to improve nitro docs.
